### PR TITLE
Implement binop sub, add debug mode and fix a bug.

### PR DESCRIPTION
**Pull request by ptersilie**
This PR does a couple of things required to get the ykbf interpreter up and running.

1) Implement the binary operation `sub`, which is very similar to `add`. Eventually this will be replaced by a macro.
2) Add a debug mode, which help us debug our compiled traces, since by default gdb/lldb can't set breakpoints into our compiled code (the reason for which we are still unsure about).
3) Fix a bug in the deref implementation.

I've not spend a great amount of time documenting some of this stuff, since most of it is gonna change significantly soon when we merge the new IR in.

### CODE REVIEW

File: ykcompile/src/lib.rs [hunk 1]
```diff
@@ -1511,8 +1551,21 @@ impl<TT> TraceCompiler<TT> {
     }
 
     /// Finish compilation and return the executable code that was assembled.
-    fn finish(self) -> dynasmrt::ExecutableBuffer {
-        self.asm.finalize().unwrap()
+    fn finish(self, debug: bool) -> dynasmrt::ExecutableBuffer {
```
[the remark this answers is not in the document]
Comment (ptersilie) [reply ykcompile/src/lib.rs:1554]: It's here: https://github.com/softdevteam/yk/pull/138/commits/dfb98d1acd162139bf4be5582c87fad629e204bb#diff-a1de3732cb5cf80102713dfb6ecbd3ea5828e5924dc5dc5d9ebe4153d2ea40a7R1572

**Comment by bjorn3** [comment]
> Add a debug mode, which help us debug our compiled traces, since by default gdb/lldb can't set breakpoints into our compiled code (the reason for which we are still unsure about).

Does `break *0xdeadbeef`/`break -a 0xdeadbeef` not work? For breaking directly on the symbol name you will have to use the gdb jit protocol including generation of dwarf debuginfo. For backtraces you need to register unwind info using `__register_frame` like so: https://github.com/bjorn3/rustc_codegen_cranelift/blob/5103a258aa806b4072c7081c6cb71a102979852e/src/debuginfo/unwind.rs#L79-L137

File: ykcompile/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vext01) [reply ykcompile/src/lib.rs:1554]: I see, so it's hardcoded and you flip it during development. fine.

**Comment by vext01** [comment]
@ptersilie perhaps fill @bjorn3 in regarding your journey into gdb fun :)

File: ykcompile/src/lib.rs [hunk 2]
```diff
@@ -1511,8 +1551,21 @@ impl<TT> TraceCompiler<TT> {
...
+        let buf = self.asm.finalize().unwrap();
+        if debug {
+            // In debug mode the memory section which contains the compiled trace is marked as
+            // writeable, which enables gdb/lldb to set breakpoints within the compiled code.
```
Comment (bjorn3) [inline ykcompile/src/lib.rs:1558]: How??? I have never had problems with setting breakpoints for cranelift-simplejit, which only enables `PROT_EXEC`.

**Comment by ptersilie** [comment]
You can read about my journey here: https://stackoverflow.com/questions/64484683/gdb-lldb-cant-access-memory-address-of-jitted-code

I'm aware that this solution is not ideal as it won't work on BSD, for example, but it was simple enough to get us going. Since I've already spent too much time on this I'm a bit reluctant to change anything else. But I will keep your comment in mind if we ever have to debug on BSD or Mac.

File: ykcompile/src/lib.rs (hunk 2 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:1558]: Indeed. How? I've yet to find an answer to that question. If you have an idea I'd be very curious to hear it. See the stackoverflow discussion for more details.

**Comment by vext01** [comment]
FWIW, I'm OK with this. We can address the W^X stuff in another PR if we can find a solution.

Ready to merge?

**Comment by bjorn3** [comment]
Strange. Ptrace write should ignore all permission checks. Are you using selinux or some other LSM (linux security module) that may block it?

**Comment by vext01** [comment]
I don't think so. It's a bog standard Debian 10 machine.

**Comment by ptersilie** [comment]
Sure, I'll just post it here for documentary purposes:
```rust
#![feature(proc_macro_hygiene)]
extern crate dynasmrt;
#[macro_use]
extern crate dynasm;

use dynasmrt::{DynasmApi, DynasmLabelApi};

use std::{io, slice, mem};
use std::io::Write;

fn main() {
    assemble(4);
}

fn assemble(i: i32) {
    let mut ops = dynasmrt::x64::Assembler::new().unwrap();
    let a: u64 = 4294967296 + 8;
    let b1: u32 = a as u32;
    let b2: u32 = (a >> 32) as u32;

    dynasm!(ops
        ; sub rsp, 8
        ; mov DWORD [rsp], b1 as i32
        ; mov DWORD [rsp+4], b2 as i32
        ; mov rax, QWORD a as i64
        ; add rsp, 8
        ; ret
    );

    let buf = ops.finalize().unwrap();

    let compiled_fn: fn() -> u64 = unsafe {
        mem::transmute(buf.ptr(dynasmrt::AssemblyOffset(0)))
    };

    let x = execute(compiled_fn);
    println!("{}", x);
}

fn execute(f: fn() -> u64) -> u64 {
    f()
}
```
Requires `dynasmrt = 0.5.2` and `dynasm = 0.5.2` and rust nightly.

File: ykcompile/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ptersilie) [reply ykcompile/src/lib.rs:1554]: Yes, I don't think it needs to be any more fancy than that for our purposes. At least for the time being.

**Comment by ptersilie** [comment]
> Ready to merge?

Yes, don't think there's anything else to do here.

**Comment by ptersilie** [comment]
Whoops. Fixing.

**Comment by ptersilie** [comment]
Fixed. Let me know if ok to squash.

**Comment by vext01** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed!

**Comment by ptersilie** [comment]
Of course I didn't format. You can tell it's getting late again.

**Comment by ptersilie** [comment]
Fixed.

**Comment by bjorn3** [comment]
I was able to single step through it just fine. I am using `GNU gdb (Debian 8.2.1-2+b3) 8.2.1`. (the default for debian)

**Comment by vext01** [comment]
As I understand, the issue is setting a breakpoint at a jitted instruction address, not stepping. Is that right lukas?

**Comment by bjorn3** [comment]
I thought the problem was stepping. Setting a breakpoint does indeed not work for me.

**Comment by ptersilie** [comment]
Yes. Initially the problem was trying to step over (`nexti`) a function call. Which is the same as setting a breakpoint after the call instruction and thus results in the same error.

**Comment by bjorn3** [comment]
I just tested cg_clif in jit mode and setting breakpoints in the jitted code works fine. This is what simplejit uses for making the memory executable the memory: https://github.com/bytecodealliance/wasmtime/blob/a1db0ff6a9d1cb4d606d9e3a41e594b0bb0bee7b/cranelift/simplejit/src/memory.rs#L196 Maybe the difference is in how it is allocated: https://github.com/bytecodealliance/wasmtime/blob/a1db0ff6a9d1cb4d606d9e3a41e594b0bb0bee7b/cranelift/simplejit/src/memory.rs#L69

**Comment by vext01** [comment]
> Requires dynasmrt = 0.5.2

That's quite old now. We should try the newest.

**Comment by ptersilie** [comment]
Actually, in yk we use `0.7`.

**Comment by vext01** [comment]
We should try with the latest version in case it has been fixed.

**Comment by ptersilie** [comment]
> I just tested cg_clif in jit mode and setting breakpoints in the jitted code works fine. This is what simplejit uses for making the memory executable the memory: https://github.com/bytecodealliance/wasmtime/blob/a1db0ff6a9d1cb4d606d9e3a41e594b0bb0bee7b/cranelift/simplejit/src/memory.rs#L196 Maybe the difference is in how it is allocated: https://github.com/bytecodealliance/wasmtime/blob/a1db0ff6a9d1cb4d606d9e3a41e594b0bb0bee7b/cranelift/simplejit/src/memory.rs#L69

Interesting. So cranelift uses different allocation functions for different systems, e.g. `MmapMut::map_anon` for selinux, and `libc::posix_memalign` for the rest. Dynasm on the other hand seems to always use `MMapMut::map_anon`. So maybe it's the anonymous flag that causes this issue?

**Comment by bjorn3** [comment]
https://yarchive.net/comp/linux/ptrace_mmap.html seems relevant.

**Comment by ptersilie** [comment]
I see. So the problem is actually `MAP_SHARED` as that would mean that breakpoints are written back to the executable. However, Linus seems to suggest that they now turn a shared page into a private page on a ptrace write, specifically because people used to be confused when they couldn't set breakpoints in shared memory. So why am I still seeing this now if it's supposedly been fixed?

**Comment by bjorn3** [comment]
Found it:

https://github.com/torvalds/linux/blob/c4cf498dc0241fa2d758dba177634268446afb06/mm/gup.c#L930-L940

> We used to let the write,force case do COW in a VM_MAYWRITE VM_SHARED !VM_WRITE vma, so ptrace could set a breakpoint in a read-only mapping of an executable, without corrupting the file (yet only when that file had been opened for writing!). Anon pages in shared mappings are surprising: now just reject it.

Changed in https://github.com/torvalds/linux/commit/cda540ace6a194850e23c79955cc2e46fd91c19a

The solution is to use `MAP_PRIVATE` instead of `MAP_SHARED`.

**Comment by ptersilie** [comment]
Hm, maybe dynasm should use https://github.com/danburkert/memmap-rs/blob/master/src/unix.rs#L99 instead then? Although I'm not sure if this does other stuff we may not want. I will chat to its dev to see what he thinks.

**Comment by CensoredUsername** [comment]
The bigger question is why memmap is passing MAP_SHARED when making an anonymous memory map. There's no reason to do so, it'll never be shared with another process.

Unfortunately it seems like memmap is no longer being actively maintained. There's a fork of it which has this issue fixed (https://crates.io/crates/mapr) but it's really new so not sure how much to trust it.

**Comment by ptersilie** [comment]
@CensoredUsername I see. We can live without it for the time being (on Linux we can debug using our hack). But eventually we probably need to fix this. But hopefully by then `mapr` has gained a larger reputation.

**Pull request merged**